### PR TITLE
Fix unbounded recipe-local file discovery

### LINKED ISSUE #537: Editing build.conf recursively scans the configured build directory and can crash VS Code with renderer OOM
**Describe the bug**

When a workspace contains both:
- a BitBake document named `build.conf` (that the extension recognizes as a Yocto file)
- a build directory of the same name, ie `build/`

editing `build.conf` causes the extension to recursively enumerate the entire `build/` directory.

`extractRecipeName("build.conf")` returns `build`, so `getLocalFoldersForRecipeUri()` calculates the recipe-local directory as `<workspace>/build`. The extension then:

1. Calls `vscode.workspace.findFiles()` with `**/*`, no exclusion, and no result limit.
2. Calls synchronous `find.dirSync()` on the same directory.

In a Yocto workspace with millions of file representing hundreds of gigs, this repeatedly made the extension host unresponsive and eventually crashed the renderer after exhausting its 4 GB V8 heap.

The standard `"**/build/**"` VS Code exclusion does not protect this search because the `RelativePattern` base is already `<workspace>/build`; returned paths are relative to that directory.

With embedded-language support enabled, this also generated `.py` documents and activated the Python extension while the extension host was blocked, resulting in an indefinitely running “Python extension loading” notification.

**To Reproduce**
1. Create a workspace resembling:
``` 
workspace/ 
├── build.conf 
├── build/<a large generated Yocto build tree> 
└── sources/<one or more .bb files>
```
2. Configure:
  "files.associations": {
    "*.conf": "bitbake"
  },
  "bitbake.pathToBuildFolder": "${workspaceFolder}/build" }
3. Open `build.conf`.
4. Make any edit to the document.
5. Observe the extension host become unresponsive while the sibling  build/  directory is recursively searched.
6. With a sufficiently large build tree, the VS Code renderer eventually crashes.

A minimal `build.conf` is sufficient; its contents are not important.

**Expected behavior**
`build.conf` should not be treated as a recipe named  build  for recipe-local  SRC_URI  file discovery.
Recipe-local file discovery should never recursively enumerate the configured BitBake build output directory. Searches should also be bounded and cancellable.

I understand that the `.conf` file association might not be intended, but in practice it has made the extension more convenient to use.

**Desktop:**
• OS: Pop!_OS 22.04 LTS, kernel 7.0.11-76070011-generic
• VS Code version: 1.128.0 when the renderer crashes occurred
• Extension version: 2.9.0
• Yocto version: 5.0.18
• Architecture: x64

**Debug logs**
The renderer exits with code 132. Crashpad dumps show that this is V8 deliberately terminating after reaching its heap limit:

```
Reached heap limit

$handleFileMatch
addMatch
handleFindMatch
$handleFileMatch
```
The dump contains retained file-search results from paths such as:

`<workspace>/build/tmp/work-shared/.../kernel-source/...`
`<workspace>/build/tmp-k3r5/work-shared/gcc-.../gcc/testsuite/...`

The renderer log reports:
```
CodeWindow: renderer process gone (reason: crashed, code: 132)
Extension host (...) is unresponsive.
UNRESPONSIVE extension host: starting to profile NOW
```

Relevant installed extensions:
```
yocto-project.yocto-bitbake@2.9.0
ms-python.python@2026.4.0
ms-python.vscode-pylance@2026.2.1
ms-python.vscode-python-envs@1.36.0
```

**Additional context**
The problematic flow appears to be:

•  server/src/server.ts :  documents.onDidChangeContent()  requests  getRecipeLocalFiles  for every changed BitBake document.
•  client/src/documentLinkProvider.ts :  getLocalFoldersForRecipeUri()  maps  build.conf  to the sibling  build/  directory.
•  findFilesAndDirs()  performs an unbounded  workspace.findFiles()  call and a synchronous recursive  find.dirSync() .

Current upstream implementation:

• https://github.com/yoctoproject/vscode-bitbake/blob/staging/server/src/server.ts
• https://github.com/yoctoproject/vscode-bitbake/blob/staging/client/src/documentLinkProvider.ts

Potential fixes:

1. Do not request recipe-local files for non-recipe documents such as  .conf .
2. Reject candidate directories equal to or contained within  bitbake.pathToBuildFolder .
3. Add a reasonable  maxResults  limit to  workspace.findFiles() .
4. Replace  find.dirSync()  with a bounded, asynchronous, cancellable traversal.
5. Ideally perform this discovery lazily when  SRC_URI  completion is actually requested rather than on every document change.

A local patch that skips the configured build folder and caps results stopped the renderer OOM. Setting:

```"bitbake.disableEmbeddedLanguagesFiles": true```

also prevented the secondary Python-extension loading loop, while leaving core BitBake functionality enabled.

### PR DESCRIPTION
**Pull request by gcomneno**
## Summary

Fix unbounded recipe-local discovery that could recursively scan a configured build directory when editing documents such as `build.conf`.

Fixes #537.

## Root cause

Recipe-local discovery was triggered eagerly from document changes and used separate traversal paths for files and directories.

The client also attempted recipe-local discovery for `.conf` documents even though recipe-local links are not applicable there.

## Changes

- add a standalone non-regression commit covering the existing recipe-local document-link behavior;
- skip recipe-local discovery for `.conf` documents before requesting links or starting a scan;
- centralize recipe-local roots and glob-pattern construction for both document links and the language client;
- replace separate file and directory traversal with one asynchronous `fast-glob.stream()` scan;
- return files and directories from the same object-mode stream using absolute paths;
- limit traversal concurrency to 4 and avoid following symbolic links;
- cancel active discovery through `CancellationToken` and `stream.destroy()`;
- avoid adding a separate timeout;
- request recipe-local files lazily only while completing `SRC_URI`;
- cache successful recipe-local discovery results;
- preserve compatibility with clients that do not support the custom request.

## Commit structure

1. `test: cover recipe-local document links`
2. `fix: bound recipe-local file discovery`
3. `optim: defer recipe-local discovery to completion`

## Validation

- ESLint on the modified client and server files;
- `npm run compile`;
- targeted document-link provider suite: 4 tests passed;
- targeted server completion suite: 27 tests passed;
- 31 targeted tests passed in total.

### CODE REVIEW

File: client/src/__tests__/unit-tests/document-link-provider.test.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import fs from 'fs'
+import os from 'os'
+import path from 'path'
+import { type CancellationToken, type TextDocument } from 'vscode'
+import { type LanguageClient } from 'vscode-languageclient/node'
+import { BitbakeDocumentLinkProvider } from '../../documentLinkProvider'
+
+jest.mock('vscode')
+
+describe('BitbakeDocumentLinkProvider', () => {
```
Comment (deribaucourt) [inline client/src/__tests__/unit-tests/document-link-provider.test.ts:20]: Before adding the test cases specifically for the edge-cases that we are addressing, we must add the missing non-regression tests. This should have been done when the feature was added, but it looks like this class skipped them!

The regression tests should be added first, in their own commit, separate for any other change. That way we will ensure that the new implementation will not break existing behavior.

I think you are quite familiar now with our unit testing environment: Mock everything outside of this class. Reuse or define common vscode mocks when needed.

The non-regression test here should address the typical execution path: links are required for a bitbake document, we mock the underlying files, and server interactions, and we check that the output of this class contains the proper links. There's a lot to mock here, so we need extra care to keep it simple and maintainable, defining clever mocking and testing boundaries..

File: client/src/__tests__/unit-tests/document-link-provider.test.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        fsPath: '/workspace/build.conf'
+      }
+    } as unknown as TextDocument
+
+    const result = await provider.provideDocumentLinks(
+      document,
+      {} as CancellationToken
+    )
+
+    expect(result).toEqual([])
```
Comment (deribaucourt) [inline client/src/__tests__/unit-tests/document-link-provider.test.ts:206]: The verification here should be that fs.promises.readdir was not called.

File: client/src/documentLinkProvider.ts [hunk 3]
```diff
@@ -9,35 +9,237 @@ import { RequestMethod, type RequestResult } from './lib/src/types/requests'
...
+    filePattern: { base: string, pattern: string },
+    maxFileResults: number | undefined,
+    token: vscode.CancellationToken | undefined,
+    timeoutMs: number
+  ): Promise<vscode.Uri[]> {
+    if (token?.isCancellationRequested === true) {
+      return []
+    }
+
+    const cancellationTokenSource = new vscode.CancellationTokenSource()
```
Comment (deribaucourt) [inline client/src/documentLinkProvider.ts]: I'm not really sure about the cancel token usage we have now. There are two sources of cancellation:
 - A timeout
 - The cancel token

The cancel token should be enough if it is properly programmed. I think it should be represented as a "cancel" button while our finding promises are getting handled. We need a token-aware find function that will cancel when the token is activated. The find function should use an async, so that the UI is not stuck, and the cancel tooltip can render. Thus, we wouldn't need to implement an additional timeout. The user may click "cancel" if a querry is too long.

The thing is, I don't think vscode provides us with a token-aware findFolders().

File: client/src/documentLinkProvider.ts [hunk 4]
```diff
@@ -9,35 +9,237 @@ import { RequestMethod, type RequestResult } from './lib/src/types/requests'
...
+        cancellationTokenSource.cancel()
+      })
+
+    let timeout: ReturnType<typeof setTimeout> | undefined
+
     try {
-      for (let i = 0; i < filePatterns.length; i++) {
-        foundFiles.push(
-          ...(await vscode.workspace.findFiles(new vscode.RelativePattern(filePatterns[i].base, filePatterns[i].pattern), undefined, maxFileResults, token))
+      const result = await Promise.race([
```
Comment (deribaucourt) [inline client/src/documentLinkProvider.ts]: If we do keep the timeout promise race pattern, it should be wrapped as a helper function. It will reduce the cognitive complexity of the function, while providing a reusable piece of code.

Comment (deribaucourt) [reply client/src/documentLinkProvider.ts]: Actually it looks like this is what settleBeforeTimeout() is for. Could you refactor the code to use it in both places?

File: client/src/documentLinkProvider.ts [hunk 5]
```diff
@@ -9,35 +9,237 @@ import { RequestMethod, type RequestResult } from './lib/src/types/requests'
...
+        timedOut = true
+        break
+      }
+
+      const currentDir = pendingDirs.pop() as string
+
+      try {
+        const entries =
+          await BitbakeDocumentLinkProvider.settleBeforeTimeout(
+            fs.promises.readdir(currentDir, {
```
Comment (deribaucourt) [inline client/src/documentLinkProvider.ts]: Iterating the workspace twice: once for files and once for directories is quite inefficient.
I did some research and I think `fast-glob`'s `stream` function is what we should use for a token-aware files+folders search function.

File: client/src/language/languageClient.ts [hunk 6]
```diff
@@ -98,6 +98,10 @@ export async function activateLanguageServer (context: ExtensionContext, bitBake
       return { foundFileUris: [], foundDirs: [] }
     }
 
+    if (path.extname(params.uri) === '.conf') {
```
Comment (deribaucourt) [inline client/src/language/languageClient.ts]: The `.conf` file check is duplicated in two places. I fear one day we want to add that kind of feature to `.conf` and we're going to forget about it. I notice there was also a duplication of the pattern logic (`pnDir`, `filesDir`).

Could you refactor that duplicated code into the common path of these functions? Feel free to rename the common function calls, or introduce an additional layer of functions to make the most sense.

File: server/src/connectionHandlers/onCompletion.ts [hunk 7]
```diff
@@ -72,8 +82,30 @@ async function getBitBakeCompletionItems (textDocumentPositionParams: TextDocume
 
     const variablesAllowedForUriCompletion = ['SRC_URI']
     const isVariableAllowedForUriCompletion = analyzer.isStringContentOfVariableAssignment(documentUri, wordPosition.line, wordPosition.character, variablesAllowedForUriCompletion)
-    const recipeLocalFiles = analyzer.getRecipeLocalFiles(documentUri)
-    if (isVariableAllowedForUriCompletion && recipeLocalFiles !== undefined) {
+
+    if (isVariableAllowedForUriCompletion) {
+      let recipeLocalFiles = analyzer.getRecipeLocalFiles(completionDocumentUri)
+
+      if (recipeLocalFiles === undefined && connection !== undefined) {
```
Comment (deribaucourt) [inline server/src/connectionHandlers/onCompletion.ts:89]: Tested the user experience here: requesting completion twice in `SRC_URI`. The behavior is the same as before, except that it is deferred to the first call. As intended, nice !

All the test cases here are very exhaustive as well.

Maybe just rename this second commit: `
fix: defer recipe-local discovery to completion` -> `
optim: defer recipe-local discovery to completion `

**Review by deribaucourt — CHANGES_REQUESTED** [review]
Thanks for rapidly taking a hold of this issue with lots of distributed changes!
I do have a few major remarks on code performance and maintainability. Please take a look.

File: client/src/__tests__/unit-tests/document-link-provider.test.ts (hunk 1 above)
Comment (gcomneno) [reply client/src/__tests__/unit-tests/document-link-provider.test.ts:20]: Implemented. The typical document-link path is now covered first in the standalone commit `7f0e2d6d test: cover recipe-local document links`. The test mocks the language client, filesystem results, and VS Code link construction, then verifies both file and directory links independently from the refactored implementation.

File: client/src/__tests__/unit-tests/document-link-provider.test.ts (hunk 2 above)
Comment (gcomneno) [reply client/src/__tests__/unit-tests/document-link-provider.test.ts:206]: Implemented. The `.conf` regression test now spies on `fs.promises.readdir` and verifies that it is not called. It also verifies that neither the language-server request nor `fast-glob.stream()` starts for `.conf` documents.

File: client/src/documentLinkProvider.ts (hunk 3 above)
Comment (gcomneno) [reply client/src/documentLinkProvider.ts]: Implemented using the cancellation token as the only cancellation mechanism. Discovery now consumes an asynchronous `fast-glob.stream()` and destroys the active stream when the token is cancelled. No additional timeout is used, and tests cover both pre-cancelled and actively cancelled scans.

File: client/src/documentLinkProvider.ts (hunk 4 above)
Comment (gcomneno) [reply client/src/documentLinkProvider.ts]: The timeout path was removed entirely in the revised implementation, so `settleBeforeTimeout()` is no longer needed. Cancellation is handled directly through the VS Code token and `stream.destroy()`.

File: client/src/documentLinkProvider.ts (hunk 5 above)
Comment (gcomneno) [reply client/src/documentLinkProvider.ts]: Implemented with one `fast-glob.stream()` call using `objectMode` and `onlyFiles: false`. Files and directories are classified from the same asynchronous stream, with concurrency limited to 4 and cancellation wired to `stream.destroy()`.

File: client/src/language/languageClient.ts (hunk 6 above)
Comment (gcomneno) [reply client/src/language/languageClient.ts]: Implemented. The `.conf` exclusion and the `${PN}` / `files` roots are centralized in `getRecipeLocalSearch()`, while glob construction is centralized in `getRecipeLocalPatterns()`. Both the document-link provider and language client now use that common path.

File: server/src/connectionHandlers/onCompletion.ts (hunk 7 above)
Comment (gcomneno) [reply server/src/connectionHandlers/onCompletion.ts:89]: Done. The server-side commit is now `23014cbb optim: defer recipe-local discovery to completion`. The implementation and exhaustive completion tests remain otherwise unchanged.

**Comment by gcomneno** [comment]
@deribaucourt I have addressed the review feedback and updated the PR history:

- added the existing-behavior regression test in its own commit;
- skipped recipe-local discovery for `.conf` documents;
- replaced the separate traversals and timeout with one cancellable `fast-glob.stream()` scan;
- centralized the recipe-local roots and pattern construction;
- renamed the server-side commit to `optim: defer recipe-local discovery to completion`;
- updated the PR description and resolved the seven review threads.

The targeted client and server suites pass (31 tests in total), along with ESLint and the full TypeScript compilation.

Could you please take another look when convenient?

File: client/src/documentLinkProvider.ts [hunk 8]
```diff
@@ -19,105 +28,210 @@ export class BitbakeDocumentLinkProvider implements vscode.DocumentLinkProvider
...
+    return {
+      roots: [
+        path.join(parentDir, extractRecipeName(uri)),
+        path.join(parentDir, 'files')
+      ]
+    }
+  }
+
+
+public static getRecipeLocalPatterns (
```
Comment (deribaucourt) [inline client/src/documentLinkProvider.ts]: Indentation level is not right here.

File: client/src/documentLinkProvider.ts [hunk 9]
```diff
@@ -19,105 +28,210 @@ export class BitbakeDocumentLinkProvider implements vscode.DocumentLinkProvider
...
+
+    if (filenames === undefined) {
+      return [`${rootPattern}/**/*`]
+    }
+
+    return filenames.map(filename => {
+      return `${rootPattern}/**/${fg.escapePath(filename)}`
+    })
+  })
+}
```
Comment (deribaucourt) [inline client/src/documentLinkProvider.ts]: Add an empty newline after this function to separate it from the next

File: client/src/documentLinkProvider.ts [hunk 10]
```diff
@@ -19,105 +28,210 @@ export class BitbakeDocumentLinkProvider implements vscode.DocumentLinkProvider
...
+
+    const isCancelled = (): boolean => {
+      return token?.isCancellationRequested === true
+    }
+
+    if (patterns.length === 0 || isCancelled()) {
+      return { foundFiles, foundDirs }
+    }
+
+    const stream = fg.stream(patterns, {
```
Comment (deribaucourt) [inline client/src/documentLinkProvider.ts:89]: The new search method is ridiculously fast compared to the previous implementation. The UI is much more reactive. Great job!

File: client/src/documentLinkProvider.ts [hunk 11]
```diff
@@ -19,105 +28,210 @@ export class BitbakeDocumentLinkProvider implements vscode.DocumentLinkProvider
...
+    }) as GlobStream
+
+    const cancellationSubscription =
+      typeof token?.onCancellationRequested === 'function'
+        ? token.onCancellationRequested(() => {
+          stream.destroy()
+        })
+        : undefined
+
+    if (isCancelled()) {
```
Comment (deribaucourt) [inline client/src/documentLinkProvider.ts]: isCancelled was called a few instructions ago, and will be checked again in the loop. I think we can remove this one.

File: client/src/documentLinkProvider.ts [hunk 12]
```diff
@@ -19,105 +28,210 @@ export class BitbakeDocumentLinkProvider implements vscode.DocumentLinkProvider
...
+            foundFiles.length < maxFileResults
+          )
+        ) {
+          foundFiles.push(vscode.Uri.file(entry.path))
+        }
       }
     } catch (error) {
-      logger.error(`An error occurred while finding directories. ${JSON.stringify(error)}`)
+      if (!isCancelled()) {
+        logger.error(
```
Comment (deribaucourt) [inline client/src/documentLinkProvider.ts]: Always print the error, even if request was not cancelled.

File: client/src/documentLinkProvider.ts [hunk 13]
```diff
@@ -19,105 +28,210 @@ export class BitbakeDocumentLinkProvider implements vscode.DocumentLinkProvider
...
+          `An error occurred while finding recipe-local entries. ${
+            JSON.stringify(error)
+          }`
+        )
+      }
+    } finally {
+      cancellationSubscription?.dispose()
```
Comment (deribaucourt) [inline client/src/documentLinkProvider.ts:147]: Also make sure to destroy the stream, if not automatically cleaned up. Another exception may have popped that wasn't the stream breaking.

**Review by deribaucourt — CHANGES_REQUESTED** [review]
Very minor remarks with the new implementation. It provides great results, thanks a lot!

**Review by deribaucourt — APPROVED** [review]

**Comment by gcomneno** [comment]
@deribaucourt Thank you for the follow-up review, for testing the responsiveness of the new search, and for merging the contribution.

All five minor remarks were addressed in `461bb3fa`:

- corrected the method indentation and spacing;
- removed the redundant cancellation check;
- errors are now always logged;
- stream cleanup is guaranteed from `finally`;
- stream destruction is centralized and idempotent.

CI run #1633 passed completely, including lint, compilation, the full test project, the standalone language-server package, and the VSIX build.

I also performed a final focused review of the resulting implementation. I did not identify any further change that should have been folded into this PR. A possible additional test for a non-cancellation stream error remains a future, non-blocking improvement so the merged change stays focused.
